include/up-cpp/client/RpcInitiator.h

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-FileType: SOURCE
+// SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef UP_CPP_CLIENT_RPCINITIATOR_H
+#define UP_CPP_CLIENT_RPCINITIATOR_H
+
+#include <uprotocol/v1/umessage.pb.h>
+#include <uprotocol/v1/ustatus.pb.h>
+#include <up-cpp/datamodel/builder/UMessage.h>
+#include <up-cpp/transport/UTransport.h>
+
+#include <future>
+#include <memory>
+#include <variant>
+
+namespace uprotocol::client {
+
+/// @brief Interface for uEntities to initiate RPC requests
+///
+/// Like all L2 client APIs, the RpcInitiator is a wrapper on top of the L1
+/// UTransport API; in this instance, it is the request-initiating half of the
+/// RPC model.
+class RpcInitiator {
+public:
+	/// @brief Constructs an initiator connected to a given transport
+	explicit RpcInitiator(std::shared_ptr<transport::UTransport> transport);
+
+	/// @brief Gets a new UMessageBuilder configured for generating RPC request
+	///        messages targeting a specific RPC method.
+	///
+	/// @param target_method The method that will be invoked when the message
+	///                      is built and sent.
+	/// @param ttl Time in milliseconds that an RPC request will remain valid
+	///            starting from when UMessageBuilder::build() is called.
+	///
+	/// @remarks This message builder can be held and reused for recurring
+	///          requests to a particular target.
+	/// @remarks Each call to this method will produce a new builder instance.
+	///
+	/// @returns A request UMessageBuilder
+	[[nodiscard]] datamodel::builder::UMessageBuilder requestBuilder(
+	    v1::UUri&& target_method, std::chrono::milliseconds ttl) const;
+
+	/// @brief Contains either a UStatus or a UMessage
+	using StatusOrMessage = std::variant<v1::UStatus, v1::UMessage>;
```

Comment:
What do you think about using this kind of alias more widely in return type?

Over in the Rust implementation we use `Result<(), UStatus>` fairly often to communicate either operation _succeed_: `()` or operation _failed_: `UStatus` with an appropriate `UCode` set for it communicate how it failed.

Perhaps it doesn't mesh as well with C++ or the chosen error model? The plan was to use exceptions if something failed?

Reply:
Any downsides to rolling our own expected type? I certainly like the idea of having a type that better expresses the semantics of what we're doing.

Reply:
Our own expected can now be found in `up-cpp/utils/Expected.h`. It copies most of the interface from `std::expected`, omitting things that are less likely to be used or that would be hard to implement.

I have included a static assert that will fire someday when we migrate to C++23 and can replace it with `std::expected`.